src/python_bindings/bindings.cpp

```diff
@@ -459,12 +463,15 @@ class PyHNSWLibIndex : public PyVecSimIndex {
     }
     PyBatchIterator createBatchIterator(const py::object &input,
                                         VecSimQueryParams *query_params) override {
+
         py::array query(input);
-        auto del = [&](VecSimBatchIterator *pyBatchIter) {
+        py::gil_scoped_release py_gil;
```

Comment:
Nice catch!
Let's also document when declaring `indexGuard` that it must be held after releasing Python GIL to avoid deadlocks

Reply:
Done!